stt.py

```diff
@@ -0,0 +1,190 @@
+#!/usr/bin/env python
+"""
+
+    Greynir: Natural language processing for Icelandic
+
+    Copyright (C) 2021 Miðeind ehf.
+
+       This program is free software: you can redistribute it and/or modify
+       it under the terms of the GNU General Public License as published by
+       the Free Software Foundation, either version 3 of the License, or
+       (at your option) any later version.
+       This program is distributed in the hope that it will be useful,
+       but WITHOUT ANY WARRANTY; without even the implied warranty of
+       MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+       GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see http://www.gnu.org/licenses/.
+
+
+    Friendly command line interface for Icelandic text to speech synthesis.
+
+"""
+
+import os
+import sys
+from urllib.request import urlopen
+
+import requests
+
+from speech import (
+    text_to_audio_url,
+    DEFAULT_VOICE,
+    SUPPORTED_VOICES,
+    DEFAULT_TEXT_FORMAT,
+    DEFAULT_AUDIO_FORMAT,
+    SUPPORTED_AUDIO_FORMATS,
+    SUPPORTED_TEXT_FORMATS,
+)
+from speech.voices import suffix4audiofmt
+from util import icelandic_asciify
+
+
+_DATA_URI_PREFIX = "data:"
+
+
+def _is_data_uri(s: str) -> bool:
+    """Returns whether a URL is a data URI (RFC2397). Tolerates uppercase prefix."""
+    return s.startswith(_DATA_URI_PREFIX) or s.startswith(_DATA_URI_PREFIX.upper())
+
+
+def _bytes4data_uri(data_uri: str) -> bytes:
+    """Returns data contained in data URI (RFC2397) as bytes."""
+    with urlopen(data_uri) as response:
+        return response.read()
+
+
+def _fetch_audio_bytes(url: str) -> bytes:
+    """Returns bytes of audio file at URL."""
+    if _is_data_uri(url):
+        return _bytes4data_uri(url)
+
+    r = requests.get(url)
+    if r.status_code != 200:
+        raise Exception(
+            f"Received HTTP status code {r.status_code} when fetching {url}"
+        )
+    return r.content
+
+
+def _play_audio_file(path: str) -> bool:
+    """Play audio file at path via command line player. This only
+    works on systems with either afplay (macOS) or mpg123 (Linux)."""
+
+    AFPLAY_PATH = "/usr/bin/afplay"
+    MPG123_PATH = "/usr/bin/mpg123"
```

Comment:
Yes, it is hacky, but automatic playback via the command line tool is a very much optional, nice-to-have sort of feature and not really required at all for anything other than minor convenience. Effectively, this is for personal testing purposes. If you don't like it, I can just get rid of the playback feature.

Reply:
No, having it is really good, but maybe you should support a combination of path prefix and program name, and try them all until one succeeds.

Reply:
Mucking around with a user's $PATH is a complex and dodgy business, in my experience, but yes, this could be made smarter, e.g. via use of "which". I'll make it happen.